Drv/Ip/IpSocket.hpp

```diff
@@ -0,0 +1,166 @@
+    SOCK_FAILED_TO_SET_SOCKET_OPTIONS = -5,  //!< Failed to configure socket
+    SOCK_INTERRUPTED_TRY_AGAIN = -6,         //!< Interrupted status for retries
+    SOCK_READ_ERROR = -7,                    //!< Failed to read socket
+    SOCK_DISCONNECTED = -8,                  //!< Failed to read socket with disconnect
+    SOCK_FAILED_TO_BIND = -9,                //!< Failed to bind to socket
+    SOCK_FAILED_TO_LISTEN = -10,             //!< Failed to listen on socket
+    SOCK_FAILED_TO_ACCEPT = -11,             //!< Failed to accept connection
+    SOCK_SEND_ERROR = -13,                   //!< Failed to send after configured retries
+};
+
+/**
+ * \brief Helper base-class for setting up Berkley sockets
+ *
+ * Certain IP headers have conflicting definitions with the m_data member of various types in fprime. TcpHelper
+ * separates the ip setup from the incoming Fw::Buffer in the primary component class preventing this collision.
+ */
+class IpSocket {
+  public:
+    IpSocket();
+    /**
+     * \brief configure the ip socket with host and transmission timeouts
+     *
+     * Configures the IP handler (Tcp, Tcp server, and Udp) to use the given hostname and port. When multiple ports are
+     * used for send/receive these settings affect the send direction (as is the case for udp). Hostname DNS translation
+     * is left up to the caller and thus hostname must be an IP address in dot-notation of the form "x.x.x.x". Port
+     * cannot be set to 0 as dynamic port assignment is not supported.
+     *
+     * \param hostname: socket uses for outgoing transmissions (and incoming when tcp). Must be of form x.x.x.x
+     * \param port: port socket uses for outgoing transmissions (and incoming when tcp). Must NOT be 0.
+     * \param send_timeout_seconds: send timeout seconds portion
+     * \param semd_timeout_microseconds: send timeout microseconds portion. Must be less than 1000000
+     * \return status of configure
+     */
+    SocketIpStatus configure(const char* hostname, const U16 port, const U32 send_timeout_seconds,
+                             const U32 send_timeout_microseconds);
+
+    /**
+     * \brief check if IP socket has previously been opened
+     *
+     * Check if this IpSocket has previously been opened. In the case of Udp this will check for outgoing transmissions
+     * and (if configured) incoming transmissions as well. This does not guarantee errors will not occur when using this
+     * socket as the remote component may have disconnected.
+     *
+     * \return true if socket is open, false otherwise
+     */
+    bool isOpened(void);
+
+    /**
+     * \brief open the IP socket for communications
+     *
+     * This will open the IP socket for communication. This method error checks and validates properties set using the
+     * `configure` method.  In the case of Udp, incoming transmissions are opened if configured via `configureRecv`. A
```

Comment:
It should be possible, but this is not a specific feature we intended to support.

Reply:
UDP will typically only be used only for sending or duplex mode.

Reply:
Yeah, I was wondering if this could be a replacement for the udp sender and udp receiver components, but since this doesn't support simplex read, we can't replace those components quite yet.

Reply:
It will be single/bidirectional for UDP sockets.